Add operator code to enable istiodRemote

linsun opened the pull request:
Allow users to configure istiodRemote from istio operator.

linsun committed:
add injection url

linsun committed:
update

linsun committed:
add istiodRemote

linsun committed:
pick up latest api

linsun committed:
more update from make gen

linsun committed:
Merge branch 'master' of https://github.com/istio/istio

linsun committed:
Revert "pick up latest api"

This reverts commit 4c8322b8d425b58b498fade2e9afe6ee2acb8564.

linsun committed:
Revert "more update from make gen"

This reverts commit 072c9031f708fcc8d6320571544b82798c93f4d2.

linsun committed:
Merge branch 'master' of https://github.com/istio/istio

linsun committed:
update api version

linsun committed:
Merge branch 'master' into master

linsun committed:
add associated operator code for istiodremote

linsun committed:
Merge branch 'master' of https://github.com/linsun/istio

linsun committed:
resolve conflict

linsun commented:
Validated the following cmd and generated the yaml look correct:

```
go run "${GOPATH}"/src/istio.io/istio/istioctl/cmd/istioctl/main.go manifest generate -f ../testingdir/istio-remote0-cluster-istiod-remote.yaml --charts manifests 
```

```
$ cat ../testingdir/istio-remote0-cluster-istiod-remote.yaml                                                       
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  hub: docker.io/linsun
  tag: latest

  components:
    base:
      enabled: true
    pilot:
      enabled: false  
    istiodRemote:
      enabled: true 
    ingressGateways:
    - name: istio-ingressgateway
      enabled: false 
  
  addonComponents:
    prometheus:
      enabled: false
  values:
    global:
      # The remote cluster's name and network name must match the values specified in the
      # mesh network configuration of the main cluster.
      multiCluster:
        clusterName: remote0 
      network: network2        
      centralIstiod: true
      
      # Replace ISTIOD_REMOTE_EP with the the value of ISTIOD_REMOTE_EP set earlier.
      remotePilotAddress: 169.x.x.x
    
    istiodRemote:
      injectionURL: https://linsun:15017/inject

    base:
      validationURL: https://linsun:15017/validate

```

istio-testing commented:
@linsun: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 9e304d3fe613909cdc423590578ab9e9a6fe4d79 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/lint_istio/12859) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

linsun committed:
format change

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 9e304d3fe613909cdc423590578ab9e9a6fe4d79 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/lint_istio/12859) | `/test lint_istio`
gencheck_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11454) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11454) | `/test gencheck_istio`
lint_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/lint_istio/12860) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11454) | `/test gencheck_istio`
lint_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/lint_istio/12860) | `/test lint_istio`
unit-tests_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13376) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

linsun committed:
more update

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 71d316206b5125a3c829a0c619a730666163c54c | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13376) | `/test unit-tests_istio`
gencheck_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11455) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11455) | `/test gencheck_istio`
unit-tests_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13377) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11455) | `/test gencheck_istio`
unit-tests_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13377) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/integ-pilot-k8s-tests_istio/12927) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

linsun committed:
make gen and fix test

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13377) | `/test unit-tests_istio`
integ-pilot-k8s-tests_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/integ-pilot-k8s-tests_istio/12927) | `/test integ-pilot-k8s-tests_istio`
gencheck_istio | 53b4d79c47757063b8fde976801fa120c83380d3 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11458) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 49c69d62146efc3f8036a3bacd14d82572f04d43 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/integ-pilot-k8s-tests_istio/12927) | `/test integ-pilot-k8s-tests_istio`
gencheck_istio | 53b4d79c47757063b8fde976801fa120c83380d3 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11458) | `/test gencheck_istio`
unit-tests_istio | 53b4d79c47757063b8fde976801fa120c83380d3 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13380) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

linsun committed:
manual copy istiod-remote chart over

John has a PR https://github.com/istio/istio/pull/23725 to do this more automatically.

istio-testing commented:
@linsun: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 53b4d79c47757063b8fde976801fa120c83380d3 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/unit-tests_istio/13380) | `/test unit-tests_istio`
gencheck_istio | 1258d49514895304e36d3110312ff41405db021d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/23793/gencheck_istio/11461) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

linsun committed:
update golden

ostromart commented:
This PR looks ok but needs a test - @linsun could you please add at least a golden input/output pair or preferably something along the lines of TestManifestGenerateGateways?
